engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala

```diff
@@ -99,53 +115,53 @@ object EngineJobHashingActor {
+  private[callcaching] case class EJHAInitialHashingResults(hashes: Set[HashResult]) extends HashResultMessage
   private[callcaching] case object CheckWhetherAllHashesAreKnown
 
   sealed trait EJHAState
   case object DeterminingHitOrMiss extends EJHAState
   case object GeneratingAllHashes extends EJHAState
 
   sealed trait EJHAResponse
-  case class CacheHit(cacheResultId: Int) extends EJHAResponse
+  case class CacheHit(cacheResultId: MetaInfoId) extends EJHAResponse
   case object CacheMiss
-  case class CallCacheHashes(hashes: List[HashResult])
+  case class CallCacheHashes(hashes: Set[HashResult])
 }
 
 /**
   * Transient data for the EJHA.
   *
   * @param possibleCacheResults The set of cache results which have matched all currently tried hashes
-  * @param cacheResultsIntersected The set of hash keys which have already been compared against the cache result table
+  * @param keysCheckedAgainstCache The set of hash keys which have already been compared against the cache result table
   * @param hashesKnown The set of all hashes calculated so far (including initial hashes)
   * @param hashesNeeded Not transient but oh-so-useful for everything else.
   */
-private[callcaching] case class EJHAData(possibleCacheResults: Set[Int],
-                                         cacheResultsIntersected: List[HashKey],
-                                         hashesKnown: List[HashResult],
-                                         hashesNeeded: Iterable[HashKey]) {
+private[callcaching] case class EJHAData(possibleCacheResults: Set[MetaInfoId],
+                                         keysCheckedAgainstCache: Set[HashKey],
+                                         hashesKnown: Set[HashResult],
+                                         hashesNeeded: Set[HashKey]) {
 
   // Manipulators
-  def intersectCacheResultIds(newCacheResults: CacheResultMatchesForHashes) = {
+  def intersectCacheResults(newCacheResults: CacheResultMatchesForHashes) = {
     val newIds = newCacheResults.cacheResultIds
-    val intersectedIds = if (cacheResultsIntersected.nonEmpty) possibleCacheResults.intersect(newIds) else newIds
+    val intersectedIds = if (keysCheckedAgainstCache.nonEmpty) possibleCacheResults.intersect(newIds) else newIds
     this.copy(
       possibleCacheResults = intersectedIds,
-      cacheResultsIntersected = cacheResultsIntersected ++ newCacheResults.hashResults.map(_.hashKey))
+      keysCheckedAgainstCache = keysCheckedAgainstCache ++ newCacheResults.hashResults.map(_.hashKey))
   }
-  def withNewKnownHashes(hashResults: Iterable[HashResult]) = this.copy(hashesKnown = hashesKnown ++ hashResults) // NB doesn't update hashesNeeded which should be considered immutable
+  def withNewKnownHashes(hashResults: Set[HashResult]) = this.copy(hashesKnown = hashesKnown ++ hashResults) // NB doesn't update hashesNeeded which should be considered immutable
```

Comment:
Oh I didn't notice the comment over there on the red side too. I remember this thing now. That comment doesn't at all help my understanding here :'(